### PR TITLE
Refactor create general order code

### PR DESCRIPTION
**Pull request by jthmcts**
### Jira link (if applicable)
https://tools.hmcts.net/jira/browse/DFR-2646

### Change description ###
Refactor create general order callback code to use handlers

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/creategeneralorder/CreateGeneralOrderAboutToStartHandler.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+                                                 IdamService idamService, Clock clock) {
+        super(finremCaseDetailsMapper);
+        this.idamService = idamService;
+        this.clock = clock;
+    }
+
+    @Override
+    public boolean canHandle(CallbackType callbackType, CaseType caseType, EventType eventType) {
+        if (!ABOUT_TO_START.equals(callbackType)) {
+            return false;
```
Comment (rishikrsharma) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/creategeneralorder/CreateGeneralOrderAboutToStartHandler.java:49]: Please move them into separate handlers. One for Contested and one for Consented.

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/creategeneralorder/CreateGeneralOrderAboutToSubmitHandler.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    public CreateGeneralOrderAboutToSubmitHandler(FinremCaseDetailsMapper finremCaseDetailsMapper,
+                                                  GeneralOrderService generalOrderService) {
+        super(finremCaseDetailsMapper);
+        this.generalOrderService = generalOrderService;
+    }
+
+    @Override
+    public boolean canHandle(CallbackType callbackType, CaseType caseType, EventType eventType) {
+        if (!ABOUT_TO_SUBMIT.equals(callbackType)) {
+            return false;
```
Comment (rishikrsharma) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/creategeneralorder/CreateGeneralOrderAboutToSubmitHandler.java:35]: same - move to separate handler

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/GeneralOrderService.java [hunk 3]
```diff
@@ -150,103 +143,57 @@ private Map<String, Object> previewGeneralOrderData(GeneralOrderPreviewDocument
...
-
-        Map<String, Object> caseData = caseDetails.getData();
+    public void addConsentedGeneralOrderToCollection(FinremCaseData caseData) {
+        GeneralOrderWrapper generalOrderWrapper = caseData.getGeneralOrderWrapper();
+        generalOrderWrapper.setGeneralOrderLatestDocument(generalOrderWrapper.getGeneralOrderPreviewDocument());
 
-        GeneralOrderConsented generalOrder =
-            new GeneralOrderConsented(documentHelper.convertToCaseDocument(caseData.get(GENERAL_ORDER_PREVIEW_DOCUMENT)),
-                getAddressToFormatted(caseData));
+        GeneralOrder generalOrder = new GeneralOrder();
```
Comment (rishikrsharma) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/GeneralOrderService.java:150]: may be use builder here

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/GeneralOrderService.java [hunk 4]
```diff
@@ -150,103 +143,57 @@ private Map<String, Object> previewGeneralOrderData(GeneralOrderPreviewDocument
...
+            }
+            generalOrderWrapper.getGeneralOrdersConsent().add(contestedGeneralOrderCollection);
         } else {
-            List<ContestedGeneralOrderCollection> generalOrderList = Optional.ofNullable(caseData.get(GENERAL_ORDER_COLLECTION_CONTESTED))
-                .map(this::convertToGeneralOrderContestedList)
-                .orElse(new ArrayList<>());
-            generalOrderList.add(generalOrderObj);
-
-            caseData.put(GENERAL_ORDER_COLLECTION_CONTESTED, generalOrderList);
+            if (generalOrderWrapper.getGeneralOrders() == null) {
```
Comment (rishikrsharma) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/GeneralOrderService.java:186]: Optional.OfNullable will be better

**Review by gmikerlemos — CHANGES_REQUESTED** [review]
Overall ok but the handler must be split and the static method in FinremCaseDetails seems to be misplaced.

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/creategeneralorder/CreateGeneralOrderAboutToSubmitHandler.java (hunk 2 above)
Comment (jthmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/creategeneralorder/CreateGeneralOrderAboutToSubmitHandler.java]: Done

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/GeneralOrderService.java (hunk 3 above)
Comment (jthmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/GeneralOrderService.java]: Done

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/creategeneralorder/CreateGeneralOrderAboutToSubmitHandler.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    @Override
+    public GenericAboutToStartOrSubmitCallbackResponse<FinremCaseData> handle(
+        FinremCallbackRequest callbackRequestWithFinremCaseDetails, String userAuthorisation) {
+
+        log.info("Create General Order about to submit callback for case id: {}",
+            callbackRequestWithFinremCaseDetails.getCaseDetails().getId());
+
+        FinremCaseDetails caseDetails = callbackRequestWithFinremCaseDetails.getCaseDetails();
+        FinremCaseData caseData = caseDetails.getData();
+        if (isContestedCase(caseDetails)) {
```
[the remark this answers is not in the document]
Comment (jthmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/creategeneralorder/CreateGeneralOrderAboutToSubmitHandler.java]: Code has been removed

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/FinremCallbackRequest.java [hunk 6]
```diff
@@ -22,4 +23,10 @@ public class FinremCallbackRequest {
 
     @JsonProperty("event_id")
     private EventType eventType;
+
+    public static FinremCallbackRequest from(FinremCaseData caseData) {
```
[the remark this answers is not in the document]
Comment (jthmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/FinremCallbackRequest.java]: Removed

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/GeneralOrderServiceTest.java [hunk 7]
```diff
@@ -146,31 +147,24 @@ public void submitContestedGeneralOrder() throws Exception {
     }
 
     @Test
-    public void generateGeneralOrderConsentedInContested() throws Exception {
-        Map<String, Object> documentMap = generalOrderService.createGeneralOrder(AUTH_TOKEN, consentedInContestedCaseDetails());
+    public void testAddContestedGeneralOrderInCollection_ConsentedInContested() throws Exception {
```
[the remark this answers is not in the document]
Comment (jthmcts) [reply src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/GeneralOrderServiceTest.java]: Refactoring has meant this method has now changed

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/creategeneralorder/CreateGeneralOrderAboutToSubmitHandler.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        } else {
+            generalOrderService.addConsentedGeneralOrderToCollection(caseData);
+        }
+
+        return GenericAboutToStartOrSubmitCallbackResponse.<FinremCaseData>builder()
+            .data(caseData)
+            .build();
+    }
+
+    private boolean isContestedCase(FinremCaseDetails caseDetails) {
```
[the remark this answers is not in the document]
Comment (jthmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/creategeneralorder/CreateGeneralOrderAboutToSubmitHandler.java]: Code has been removed

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/GeneralOrderService.java (hunk 4 above)
Comment (jthmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/GeneralOrderService.java]: Done

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/GeneralOrderService.java [hunk 9]
```diff
@@ -150,103 +143,57 @@ private Map<String, Object> previewGeneralOrderData(GeneralOrderPreviewDocument
...
-        GeneralOrderConsentedData consentedData = new GeneralOrderConsentedData(UUID.randomUUID().toString(), generalOrder);
+        GeneralOrderCollectionItem item = new GeneralOrderCollectionItem();
+        item.setId(UUID.randomUUID().toString());
+        item.setGeneralOrder(generalOrder);
 
-        List<GeneralOrderConsentedData> generalOrderList = Optional.ofNullable(caseData.get(GENERAL_ORDER_COLLECTION_CONSENTED))
-            .map(this::convertToGeneralOrderConsentedList)
-            .orElse(new ArrayList<>());
-        generalOrderList.add(consentedData);
+        if (generalOrderWrapper.getGeneralOrderCollection() == null) {
```
[the remark this answers is not in the document]
Comment (jthmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/GeneralOrderService.java]: Done

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/correspondence/FinremEmailOnlyAllSolicitorsCorresponder.java [hunk 10]
```diff
@@ -34,6 +35,11 @@ protected boolean shouldSendRespondentSolicitorEmail(FinremCaseDetails caseDetai
         return notificationService.isRespondentSolicitorDigitalAndEmailPopulated(caseDetails);
     }
 
+    protected boolean shouldSendIntervenerSolicitorEmail(FinremCaseDetails caseDetails,
```
[the remark this answers is not in the document]
Comment (jthmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/correspondence/FinremEmailOnlyAllSolicitorsCorresponder.java]: Done

**Pull request merged**